docs/design/2021-09-13-parser-as-submodule-of-tidb.md

```diff
@@ -0,0 +1,89 @@
+
+Problems above have defeated the previous argument `only go.mod need to change accordingly`. We only separated code into two repositories, actual work did not decrease. However, our development experience does getting worse.
+
+## Detailed Plan
+
+Obviously, we could move back to the old good way. But it must be a go sub-module, because:
+
+1. To keep compatibility. Parser has been imported by other tools and other users.
+2. To keep the clean dependency. That is why it is separated. We don't want to make things worse again.
+
+The detailed plan:
+
+1. Create parser directory, a sub-module, in pingcap/tidb, which is synchronized with pingcap/parser, with `replace github.com/pingcap/parser => ./parser`.
+2. Stop sending new PRs and issues to pingcap/parser, new PRs should be directly sent to pingcap/tidb. And merge recent PRs in pingcap/parser as much as possible. Changes will be synchronized from pingcap/parser to pingcap/tidb(manually), of course.
+3. Announce the deprecation of pingcap/parser to possible users. Clean up old issues and PRs, create it in pingcap/tidb again, close it, or just forget about it.
+4. We can create wrapper packages for pingcap/parser, which re-exports things from `pingcap/tidb/parser`. I mean something like `import ( . github.com/pingcap/tidb/parser/xxx)`, create a dummy function will eliminate the error of `not used import`. All code of pingcap/parser will be removed in this step. This will delevery new updates without the need of migrating importing paths.
+5. After another one or two dev cycles, we could archive pingcap/parser. It mainly depends on users of pingcap/parser.
+
+There will be a tracking issue, and the whole progress is public. Step 1 or 2 will likely take one or two weeks. Internally we could do the migration for internal tools while doing step 3.
+
+Step 4 and 5 are long-running tasks to smoothly end the support of the old parser.
+
+## Questions
+
+- What about projects importing parser?
+
+Nothing needs to be done before step 5. However, one will need to migrate the import path eventually, before the final deprecation.
+
+- What about some complex dependency chain? For example, `pkg -> forked tidb -> other pingcap/parser`.
+
+Nothing different for the complex dependencies. Just do the migration before the deprecation. And we don't actually remove the old commits for not breaking old apps.
+
+- What about CI and tests?
+
+It is not a problem. pingcap/parser only has unit tests, and integration tests `make gotest` of `pingcap/tidb`. We could run integration tests of tidb, and only run unit tests.
+
+- How about release and versioning?
+
+Since sub-module is merely another module inside another module, we need to release it separately from the main module. That means our release team needs to tag `parser/vX.X.X`, which can be solved by a script: thanks to the identical release cycle/version between parser and tidb.
```

Comment:
or we could use true semver for the parser sub-module instead of tying with the product release cycle

Reply:
I would like to leave that as another RFC.

Reply:
And it is probably a good idea to keep identical version with tidb, before it becomes a real "standalone" parser.

Reply:
Yes, IMHO semver is a bigger topic than this, considering all components of TiDB & other components.